### DISCUSSION TITLE
lifetimes hell (with UserData-s)

### DISCUSSION THREAD

**Discussion by ildar** [mlua-rs/mlua#123] [Q&A] [answered]
Hello,
I'm doing Lua module representing `calloop` crate.
I found that code is spoiled with lifetime annotations very quickly. Example:
``` Rust
// TimerHandle
#[derive(Clone)]
struct TimerHandle<'a> (rs_timer::TimerHandle<LuaValue<'a>>);
impl LuaUserData for TimerHandle<'_> {}
impl<'a> TimerHandle<'a> {
    fn new(lua: &'static Lua, ls: rs_timer::TimerHandle<LuaValue<'static>>) -> LuaResult<LuaTable<'a>> {
        let res = lua.create_table()?;
        res.set("_self",
            lua.create_userdata(
                TimerHandle(ls)
            )?)?;
        res.set("add_timeout", lua.create_function(TimerHandle::add_timeout)?)?;
        Ok(res)
    }

    pub fn add_timeout(_: &Lua, (luaself, delay_from_now, data): (LuaTable, u64, LuaValue<'static>)) -> LuaResult<()> {
        let ud: LuaAnyUserData = luaself.get("_self")?;
        let ref_self = ud.borrow::<TimerHandle<'static>>()?;
        ref_self.0.add_timeout(std::time::Duration::from_millis(delay_from_now), data);
        Ok(())
    }
}
```
I may be very wrong but think that it's because `AnyUserData` is `'static` albeit `'lua` surely would be enough.

**Q**: is it very difficult to have an alternative `AltAnyUserData` with `'lua` lifetime?

**Comment by khvzak** [comment, marked as answer]
I would suggest to try to store `LuaValue<'a>` as an associated value to a UserData (see [`set_user_value`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html#method.set_user_value) and [`get_user_value`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html#method.get_user_value))

Also you could try to store `LuaValue` in Lua Registry and refer to it via [`RegistryKey`](https://docs.rs/mlua/latest/mlua/struct.RegistryKey.html).
It would significantly simplity the "lifetimes hell".

Also I'm working on alternative implementation that will not require lifetimes for values but it's mostly a concept now.
